### PR TITLE
stale-while-revalidate implementation update

### PR DESCRIPTION
**Pull request by hthetiot**
For Reviewer
- https://github.com/kaazing/http2-cache.js/pull/66/files#diff-63fdc6a01d0ec0cbf9772723f2356e46R211
- https://github.com/kaazing/http2-cache.js/pull/66/files#diff-3891df25a9a726b84ac6e031b78b8ef9R257
- https://github.com/kaazing/http2-cache.js/pull/66/files#diff-3891df25a9a726b84ac6e031b78b8ef9R289

### CODE REVIEW

File: lib/xhr.js [hunk 1]
```diff
@@ -189,157 +179,177 @@ function enableXHROverH2(XMLHttpRequest, configuration) {
...
+                    self._changeState(XMLHttpRequest.HEADERS_RECEIVED, {'response': response});
+
+                    response.on('data', function (data) {
+                        response.data = mergeTypedArrays(response.data, data);
+                        self._changeState(XMLHttpRequest.LOADING, {'response': response});
+                    });
+
+                    response.on('finish', function () {
+
+                        cache.validated(requestInfo);
```
Comment (hthetiot) [inline lib/xhr.js]: Clear revalidate outside of cache.

File: lib/xhr.js [hunk 2]
```diff
@@ -189,157 +179,177 @@ function enableXHROverH2(XMLHttpRequest, configuration) {
...
 
-                    // Export cached response
+                    // only other option in spec is a String
+                    // inject content-length TODO remove this as should not be required
+                    requestHeaders.set('content-length', body.toString().length);
+                }
+
+
+
+                cache.revalidate(requestInfo);
```
Comment (hthetiot) [inline lib/xhr.js]: Start revalidate

**Comment by dpwspoon** [comment]
I would expect a call to `cache.revalidate(requestInfo);` on https://github.com/kaazing/http2-cache.js/pull/66/files#diff-76c4bbb65fb81fc4fc6a813da39add7bR122,  And a call to validated https://github.com/kaazing/http2-cache.js/pull/66/files#diff-76c4bbb65fb81fc4fc6a813da39add7bR136.

I also think you want to remove the entry in is Validating on in case of errors.  I.E.  https://github.com/kaazing/http2-cache.js/pull/66/files#diff-76c4bbb65fb81fc4fc6a813da39add7bR144 and maybe here: https://github.com/kaazing/http2-cache.js/pull/66/files#diff-3891df25a9a726b84ac6e031b78b8ef9R309

The expected test case for this would be:  User initiated request to http://example.com, push promise request with cache-control:no-cache,  response with stale-while-revalidating=large.  Then another user-initiated request to  http://example.com that is served by cache because stale-while-revalidating and it is validating (because push promise is in flight).  Then response to push promise completes.

File: lib/configuration.js [hunk 3]
```diff
@@ -118,35 +116,45 @@ confProto.getTransportUrl = function (url) {
...
+    // TODO pass pushRequest for future dedup pending requestInfo
+    cache.revalidate(requestInfo);
+
     pushRequest.on('response', function (response) {
+
         // TODO match or partial move to _sendViaHttp2 xhr.js to avoid maintain both
         response.on('data', function (data) {
             response.data = mergeTypedArrays(response.data, data);
         });
-
```
Comment (hthetiot) [inline lib/configuration.js]: Moved to clean on L130

File: lib/xhr.js [hunk 4]
```diff
@@ -189,157 +179,177 @@ function enableXHROverH2(XMLHttpRequest, configuration) {
...
                     }
 
-                    // Export cached response
+                    // only other option in spec is a String
+                    // inject content-length TODO remove this as should not be required
+                    requestHeaders.set('content-length', body.toString().length);
+                }
+
+                // TODO use isRevalidating and store request in cache.revalidate 
+                // cache.isRevalidating(requestInfo)
```
Comment (hthetiot) [inline lib/xhr.js]: Future pending request changes here

**Pull request merged**